lib/smack/DSAWrapper.cpp

```diff
@@ -215,23 +118,115 @@ unsigned DSAWrapper::getPointedTypeSize(const Value *v) {
 }
 
 int DSAWrapper::getOffset(const Value *v) {
-  return getOffset(new MemoryLocation(v));
+  if (DG->hasCell(*v)) {
+    auto cell = DG->getCell(*v);
+    auto node = cell.getNode();
+    // Be consistent with the old implementation.
+    if (node->isOffsetCollapsed())
+      return -1;
+    auto offset = cell.getOffset();
+    assert(offset <= INT_MAX && "Cannot handle large offsets");
+    return offset;
+  }
+  // Return 0 if we can't find the cell that `v` associates with.
+  return 0;
+}
+
+const sea_dsa::Node *DSAWrapper::getNode(const Value *v) {
+  // For sea-dsa, a node is obtained by getting the cell first.
+  if (DG->hasCell(*v)) {
+    auto node = DG->getCell(*v).getNode();
+    assert(node && "DSNode should not be NULL.");
+    return node;
+  }
+  llvm_unreachable("Values should have cells.");
 }
 
-// TODO: Should this return the node or its leader?
-const DSNode *DSAWrapper::getNode(const Value *v) {
-  const llvm::EquivalenceClasses<const llvm::DSNode *> &eqs =
-      nodeEqs->getEquivalenceClasses();
-  auto *N = nodeEqs->getMemberForValue(v);
-  return N ? eqs.getLeaderValue(N) : nullptr;
+bool DSAWrapper::isTypeSafe(const Value *v) {
+  typedef std::unordered_map<unsigned, bool> FieldMap;
+  typedef std::unordered_map<const sea_dsa::Node *, FieldMap> NodeMap;
+  static NodeMap nodeMap;
+
+  auto node = getNode(v);
+
+  if (node->isOffsetCollapsed() || node->isExternal() || node->isIncomplete() ||
+      node->isUnknown() || node->isIntToPtr() || node->isPtrToInt())
+    // We consider it type-unsafe to be safe for these cases
+    return false;
+
+  if (!nodeMap.count(node)) {
+    // Iterate all the fields of a node to find out
+    // the type-safety of each field. Then, we cache the results.
+    FieldMap fieldMap;
+    auto &types = node->types();
+    std::set<unsigned> offsets;
+
+    for (auto &t : types)
+      offsets.insert(t.first);
+
+    auto offsetIterator = offsets.begin();
+
+    while (true) {
+      if (offsetIterator == offsets.end())
+        // We have reached the last field and exit the loop
+        break;
+
+      unsigned offset = *offsetIterator;
+
+      auto &typeSet = types.find(offset)->second;
+
+      auto ti = typeSet.begin();
+      if (++ti != typeSet.end())
+        // If there are multiple access types, then it's trivially type-unsafe.
+        fieldMap[offset] = false;
+
+      // Get the maximum length
+      unsigned fieldLength = 0;
+      for (auto &t : typeSet) {
+        // TODO: fix the const_cast
+        unsigned length =
+            dataLayout->getTypeStoreSize(const_cast<llvm::Type *>(t));
+        if (length > fieldLength)
+          fieldLength = length;
+      }
+
+      // Check if the current field overlaps with the next *fields*
+      for (auto oi = ++offsetIterator; oi != offsets.end(); ++oi) {
+        bool overlap = false;
```

Comment:
Why do you need the `overlap` local variable?
To me it seems that breaking directly from the if condition would be semantically equivalent (by simpler) to what you have right now. No?

Reply:
Yes, you're right. It should break the loop from the else branch.